### PR TITLE
[Reviewer: Steve] Sfr487952

### PR DESCRIPTION
**Pull request by Tom-Rees**
Fix for http://sfr/prodAsp/scripts/MSG/Issue.asp?issue_id=439117.

Testing on l3-cc4-hstead1 now.

### CODE REVIEW

**Comment by Tom-Rees** [comment]
Hi Steve,

Please can you review these changes?

The slightly controversial bit is removing the 'display_user' call from 'create_user'. This is to avoid dumping out the iFC and covering up any error messages (User already exists).
The problem is that we can't tell between this case and the 'mainline' case, and so either must print the iFC on both, or neither. Our other tools generally display user on success and so I'm not too happy about going against that. But at the same time, create_user is the one script where you see exactly what is being put into the user, so it's probably fine.

Does that sound ok?

File: src/metaswitch/ellis/prov_tools/create_user.py [hunk 1]
```diff
@@ -70,7 +70,7 @@ def main():
         private_id = "%s@%s" % (dn, args.domain)
 
         if utils.create_user(private_id, public_id, args.domain, args.password, ifc, plaintext=args.plaintext):
-            if not args.quiet and not utils.display_user(public_id):
+            if not args.quiet:
```
Comment (steelyvoid) [inline src/metaswitch/ellis/prov_tools/create_user.py]: Doesn't this mean that success will always be set to false, unless args.quiet was set?  Presumably, the old code used display_user's returncode to confirm that the user was successfully readable.

File: src/metaswitch/ellis/prov_tools/utils.py [hunk 2]
```diff
@@ -241,8 +244,13 @@ def display_user(public_id, short=False):
             print ifc_str
         else:
             _log.error("Failed to retrieve iFC for public ID %s - HTTP status code %d", public_id, response.code)
+            if response.code == 404:
+                user_missing = user_missing and True
```
Comment (steelyvoid) [inline src/metaswitch/ellis/prov_tools/utils.py]: What is this line for?  Surely "blah = blah and true" is a noop?  Is the intention only to have user_missing if you get a 404 from both the private ID and iFC lookup (or just private ID if you're doing a "short" display)?  If so, then presumably you want to set user_missing to true at the top and set it to false in the various branches unless a 404 is encountered - is that right?

**Comment by steelyvoid** [comment]
I guess if there's no way to tell from the create whether the user was present, then we could live with the restriction you mention.  However, the display_user call does at least have the merit of checking that the user just created is readable.  Could the display_user call be refactored in such a way that it can optionally do the "readable" check without displaying anything?  Then you could call it in this mode from create_user before doing anything and determine inline whether a create is likely to hit a "user already configured" error.

**Comment by Tom-Rees** [comment]
Hi Steve,

I've updated display user to have a quiet keyword and fixed the user_missing fail.

Please could you review?

File: src/metaswitch/ellis/prov_tools/utils.py [hunk 3]
```diff
@@ -227,6 +228,8 @@ def display_user(public_id, short=False):
                 success = False
     else:
         _log.error("Failed to retrieve private IDs for public ID %s - HTTP status code %d", public_id, response.code)
+        if response.code != 404:
+            user_missing = False
```
Comment (steelyvoid) [inline src/metaswitch/ellis/prov_tools/utils.py]: As discussed, this still needs work.  As you suggested, best to have a separate missing flag for the two senses in which the user data can be missing, and only display "user missing" if they're both missing, or if we're doing the short display and the first check hit missing

File: src/metaswitch/ellis/prov_tools/utils.py [hunk 4]
```diff
@@ -197,11 +197,12 @@ def delete_user(private_id, public_id, force=False):
 
     return success
 
-def display_user(public_id, short=False):
+def display_user(public_id, short=False, quiet=False):
```
Comment (steelyvoid) [inline src/metaswitch/ellis/prov_tools/utils.py]: I think we should probably have a comment here to explain why we have a "quiet" flag in a function whose apparent purpose is to display stuff (e.g. "Set 'quiet' true to check user for readability without displaying user contents")

**Pull request merged**